internal/server/middleware.go

```diff
@@ -59,36 +59,59 @@ func handleInfraDestinationHeader(rCtx access.RequestContext, uniqueID string) e
 	return nil
 }
 
-// authenticateRequest is call for requests to routes that require authentication.
-// It validates the access key and organization, updates the lastSeenAt of the user,
+// authenticateRequest authenticates the user performing the request.
+//
+// If the route requires authentication, authenticateRequest validates the access
+// key and organization, updates the lastSeenAt of the user,
 // and may also update the lastSeenAt of the destination if the appropriate header
 // is set.
-// authenticateRequest is also responsible for adding RequestContext to the
-// gin.Context.
-// See validateRequestOrganization for a related function used for unauthenticated
-// routes.
-func authenticateRequest(c *gin.Context, srv *Server) (access.Authenticated, error) {
+//
+// If the route does not require authentication, authenticateRequest will attempt
+// the same authentication, but ignore the error. It will also look up the
+// organization from the domain name when no access key is provided.
+func authenticateRequest(c *gin.Context, route routeSettings, srv *Server) (access.Authenticated, error) {
 	tx, err := srv.db.Begin(c.Request.Context())
 	if err != nil {
 		return access.Authenticated{}, err
 	}
 	defer logError(tx.Rollback, "failed to rollback middleware transaction")
 
 	authned, err := requireAccessKey(c, tx, srv)
-	if err != nil {
+	if !route.authenticationOptional && err != nil {
 		return authned, err
 	}
 
-	if _, err := validateOrgMatchesRequest(c.Request, tx, authned.Organization); err != nil {
+	org, err := validateOrgMatchesRequest(c.Request, tx, authned.Organization)
+	if err != nil {
 		logging.L.Warn().Err(err).Msg("org validation failed")
 		return authned, internal.ErrBadRequest
 	}
-	tx = tx.WithOrgID(authned.Organization.ID)
 
-	if uniqueID := c.Request.Header.Get("Infra-Destination"); uniqueID != "" {
-		rCtx := access.RequestContext{DBTxn: tx, Authenticated: authned}
-		if err := handleInfraDestinationHeader(rCtx, uniqueID); err != nil {
-			return authned, err
+	if route.authenticationOptional {
+		// See this diagram for more details about this request flow
+		// when an org is not specified.
+		// https://github.com/infrahq/infra/blob/main/docs/dev/organization-request-flow.md
+
+		// TODO: use an explicit setting for this, don't overload EnableSignup
+		if org == nil && !srv.options.EnableSignup { // is single tenant
```

Comment:
I've been generally hesitant to go down the "single-tenant" path versus everything being multi-tenant but with sign-up disabled. Is there anything stopping us from assuming the default org in any case the org in nil? Seems correct to me, then we can move to request failure once existing sessions for single-tenant deployments have expired.